source/ShopPanel.cpp

```diff
@@ -222,81 +222,212 @@ void ShopPanel::CheckForMissions(Mission::Location location)
 
 
 
-void ShopPanel::FailSell(bool toStorage) const
+ShopPanel::TransactionResult ShopPanel::CanBuyToCargo() const
 {
+	return false;
 }
 
 
 
-bool ShopPanel::CanSellMultiple() const
+ShopPanel::TransactionResult ShopPanel::CanBuy() const
 {
-	return true;
+	return false;
 }
 
 
 
-// Helper function for UI buttons to determine if the selected item is
-// already owned. Affects if "Install" is shown for already owned items
-// or if "Buy" is shown for items not yet owned.
-//
-// If we are buying into cargo, then items in cargo don't count as already
-// owned, but they count as "already installed" in cargo.
-bool ShopPanel::IsAlreadyOwned() const
+void ShopPanel::BuyToCargo()
 {
-	return (playerShip && selectedOutfit && player.Cargo().Get(selectedOutfit))
-		|| player.Storage().Get(selectedOutfit);
 }
 
 
 
-bool ShopPanel::ShouldHighlight(const Ship *ship)
+void ShopPanel::Buy()
 {
-	return (hoverButton == 's');
 }
 
 
 
-void ShopPanel::DrawKey()
+void ShopPanel::Sell(bool /* storeOutfits */)
 {
 }
 
 
 
-int ShopPanel::VisibilityCheckboxesSize() const
+void ShopPanel::Sell()
 {
-	return 0;
 }
 
 
 
-void ShopPanel::ToggleForSale()
+ShopPanel::TransactionResult ShopPanel::CanInstall() const
 {
-	CheckSelection();
-	delayedAutoScroll = true;
+	return false;
 }
 
 
 
-void ShopPanel::ToggleStorage()
+void ShopPanel::Install()
 {
-	CheckSelection();
-	delayedAutoScroll = true;
 }
 
 
 
-void ShopPanel::ToggleCargo()
+ShopPanel::TransactionResult ShopPanel::CanUninstall() const
+{
+	return false;
+}
+
+
+
+void ShopPanel::Uninstall()
 {
-	CheckSelection();
-	delayedAutoScroll = true;
+}
+
+
+
+bool ShopPanel::CanMoveToCargo() const
+{
+	return false;
+}
+
+
+
+void ShopPanel::MoveToCargo()
+{
+}
+
+
+
+ShopPanel::TransactionResult ShopPanel::CanMoveToStorage() const
+{
+	return false;
+}
+
+
+
+void ShopPanel::MoveToStorage()
+{
+}
+
+
+
+bool ShopPanel::CanSellMultiple() const
+{
+	return true;
+}
+
+
+
+ShopPanel::TransactionResult ShopPanel::CanDoTransaction(const SDL_Keycode key) const
+{
+	switch(key)
+	{
+		case 'b': // _Buy
+			return CanBuy();
+		case 's': // _Sell
+			return CanSell();
+		case 'r': // Sell Ship and put Outfits in Storage / Move to Storage
+			if(isOutfitter)
+				return CanMoveToStorage();
+			// Shipyard, sell Ship, send Cargo to Storage
+			return CanSell();
+		default:
+			break;
+	}
+
+	if(isOutfitter)
+	{
+		switch(key)
+		{
+			case 'c': // Move to/Buy to _Cargo
+			{
+				const auto canMove = CanMoveToCargo();
+				const auto canBuy = CanBuyToCargo();
+				if(canMove || canBuy)
+					return true;
+				if(!canBuy)
+					return canBuy; // Return why we cannot buy
```

Comment:
```suggestion
					// Return why we cannot buy.
					return canBuy;
```

Reply:
Yeah, I'm going to revisit all the comments, don't feel like you have to look at them all right now.